cdr/resolver.go

```diff
@@ -0,0 +1,247 @@
+		}
+		return nil, fmt.Errorf("Query failed wtih %s", dns.RcodeToString[respObj.Status])
+	}
+
+	return parseAnswer(respObj.Answer)
+}
+
+type queryResult struct {
+	records []*dns.CAA
+	err     error
+}
+
+type caaSet []*dns.CAA
+
+func (cs caaSet) Len() int           { return len(cs) }
+func (cs caaSet) Less(i, j int) bool { return cs[i].Value < cs[j].Value } // sort by value...?
+func (cs caaSet) Swap(i, j int)      { cs[i], cs[j] = cs[j], cs[i] }
+
+func marshalCanonicalCAASet(set []*dns.CAA) ([]byte, error) {
+	var err error
+	offset, size := 0, 0
+	sortedSet := caaSet(set)
+	sort.Sort(sortedSet)
+	for _, rr := range sortedSet {
+		size += dns.Len(rr)
+	}
+	tbh := make([]byte, size)
+	for _, rr := range sortedSet {
+		ttl := rr.Hdr.Ttl
+		rr.Hdr.Ttl = 0 // only variable that should jitter
+		offset, err = dns.PackRR(rr, tbh, offset, nil, false)
+		if err != nil {
+			return nil, err
+		}
+		rr.Hdr.Ttl = ttl
+	}
+	return tbh, nil
+}
+
+// LookupCAA performs a multipath CAA DNS lookup using GPDNS
+func (cdr *CAADistributedResolver) LookupCAA(ctx context.Context, domain string) ([]*dns.CAA, error) {
+	query := make(url.Values)
+	query.Add("name", domain)
+	query.Add("type", strconv.Itoa(int(dns.TypeCAA))) // CAA
+	rawQuery := query.Encode()
+
+	// min of ctx deadline and time.Now().Add(cdr.timeout)
+	caaCtx, cancel := context.WithTimeout(ctx, cdr.timeout)
+	defer cancel()
+	results := make(chan queryResult, len(cdr.Clients))
+	for addr, interfaceClient := range cdr.Clients {
+		req, err := http.NewRequest("GET", cdr.URI, nil)
+		if err != nil {
+			return nil, err
+		}
+		req.URL.RawQuery = rawQuery
+		go func(ic *http.Client, ia string) {
+			started := time.Now()
+			records, err := cdr.queryCAA(caaCtx, req, ic)
+			cdr.stats.TimingDuration(fmt.Sprintf("CDR.GPDNS.Latency.%s", ia), time.Since(started))
+			if err != nil {
+				cdr.stats.Inc(fmt.Sprintf("CDR.GPDNS.Failures.%s", ia), 1)
+				cdr.logger.Err(fmt.Sprintf("queryCAA failed [via %s]: %s", ia, err))
+			}
+			results <- queryResult{records, err}
+		}(interfaceClient, addr)
+	}
+	// collect everything
+	failed := 0
+	var CAAs []*dns.CAA
+	var canonicalSet []byte
+	var err error
+	for i := 0; i < len(cdr.Clients); i++ {
+		r := <-results
+		if r.err != nil {
+			failed++
+			if failed > cdr.maxFailures {
+				cdr.stats.Inc("CDR.QuorumFailed", 1)
+				cdr.logger.Err(fmt.Sprintf("%d out of %d CAA queries failed", len(cdr.Clients), failed))
+				return nil, r.err
+			}
+		}
+		if CAAs == nil {
+			CAAs = r.records
+			canonicalSet, err = marshalCanonicalCAASet(CAAs)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			thisSet, err := marshalCanonicalCAASet(r.records)
+			if err != nil {
+				return nil, err
+			}
+			if len(r.records) != len(CAAs) || !bytes.Equal(thisSet, canonicalSet) {
+				cdr.stats.Inc("CDR.MismatchedSet", 1)
+				return nil, errors.New("mismatching CAA record sets were returned")
```

Comment:
Nit: We may end up adding some additional debugging here, like what proxy gave the weirdness. I'm not sure if that falls into the audit log requirements or not.